tasks/Task.cpp

```diff
@@ -23,24 +21,37 @@ struct controldev_websocket::JoystickHandler : WebSocket::Handler {
     Json::Value msg;
     Json::FastWriter fast;
     Statistics statistic;
+    bool is_controlling;
 
     void onConnect(WebSocket *socket) override{
         if (connection != nullptr){
             connection->close();
         }
         connection = socket;
+        is_controlling = false;
     }
     void onData(WebSocket *socket, const char *data) override{
         msg["result"] = false;
         if (connection == socket){
-            ++task->received;
-            msg["result"] = Json::Value(task->updateRawCommand(data));
-            if (msg["result"].asBool() && task->controlling){
-                task->raw_cmd_obj.time = base::Time::now();
-                task->_raw_command.write(task->raw_cmd_obj);
-            }
-            if (!msg["result"].asBool()){
-                ++task->errors;
+            if (is_controlling){
+                ++task->received;
+                msg["result"] = task->updateRawCommand(data);
+                if (msg["result"].asBool()){
+                    task->raw_cmd_obj.time = base::Time::now();
+                    task->_raw_command.write(task->raw_cmd_obj);
+                } else {
+                    ++task->errors;
+                }
+            } else {
+                Json::Value test_msg;
+                if (task->getTestMessage(data, test_msg)){
+                    auto str = fast.write(test_msg);
```

Comment:
This smells _so bad_.

Refactor the whole thing to have `updateRawCommand` get a Json::Value object. Then, pass it *already decoded* in both codepaths. Factor out common code instead of strong-arming the existing code into new uses.

Also, there are too many levels of ifs in this method. Need to refactor to make the flow easier to follow. For instance (no error handling included):

~~~ cpp
if (connection == socket) {
   task->handleIncomingWebsocketMessage(data)
}
...

/**
 * Here, I'm not really sure of the value of WrapperJSON. I have the impression that it is only there for the purpose of getValue ?
 * I'd say, move the builder and JSON value out of WrapperJSON, passing only a const reference to the JSON value in the constructor instead. And move them to Task
 */

bool Task::handleIncomingWebsocketMessage(char const* data) {
    if (!parseJSONMessage(data, &jdata)) {
       return false;
    }

    if (is_controlling) {
       validateControlMessage(jvalue);
       parseControlMessage(jvalue, raw_cmd_obj);
       _raw_command.write(raw_cmd_obj);
    }
    else {
       validateAskControlMessage(jvalue);
    }
}

bool Task::validateAskControlMessage(JSON::Value const& value) {
    if (!wrapper->jdata.isMember("test_message")) {
        return false;
    }
    return validateControlMessage(jdata["test_message"]);
}

void Task::updateRawCommand(JSON::Value const& value) {
   WrapperJSON wrapper(value); // just gives us easy-to-use access to the JSON fields.
}
~~~

Reply:
I got your point, but I think is not so simple. I should have to verify after `validateControlMessage(jvalue)` if the message is valid to continue and after `parseControlMessage(jvalue, raw_cmd_obj)` to see if it was parsed.

Also, when doing a handshake the `parseControlMessage` should be called.

I will try to do the closest and cleanly possible to it.